src/webgldrawer.js

```diff
@@ -802,7 +841,19 @@
             let tile = event.tile;
             let tiledImage = event.tiledImage;
             let tileContext = tile.getCanvasContext();
-            let canvas = tileContext.canvas;
+            let canvas = tileContext && tileContext.canvas;
+            // if the tile doesn't provide a canvas, or is tainted by cross-origin
+            // data, marked the TiledImage as tainted so the canvas drawer can be
+            // used instead, and return immediately - data cannot be uploaded to webgl
+            if(!canvas || $.isCanvasTainted(canvas)){
```

Comment:
This is just where `isTainted` is set on the `TiledImage`. This is the first place the drawer accesses data from a loaded tile, so I think this is the logical place to test for tainted status. Once a single tile is found to be tainted, the whole `TiledImage` is marked as tainted, and thus it will always be skipped by `WebGLDrawer` and drawn instead by `CanvasDrawer`.

An alternative would be to test each tile that makes up a frame, every frame... but that seemed too expensive.

Reply:
I feel like the case where only *some* tiles for a given `TiledImage` are tainted is a pretty unlikely edge case. However, I added a guard against loading tiles from tainted `TiledImage`s just in case. If those tiles are shared by multiple `TiledImage`s, and some of them aren't tainted for some reason (super-duper edge case) they will be loaded as soon as necessary for drawing the non-tainted image.

Unloading existing non-tainted tiles should be left to the time they are unloaded from memory, because of the shared-tile edge case.